bu-liaison-inquiry.php

```diff
@@ -1,355 +1,128 @@
-		$phone_fields = explode( ',', $phone_fields );
-		unset( $_POST['phone_fields'] );
-
-		$post_vars = $this->prepare_form_post( $_POST, $phone_fields );
-
-		// Set the API Key from the site options.
-		$post_vars['IQS-API-KEY'] = $options['APIKey'];
-
-		// Setup arguments for the external API call.
-		$post_args = array( 'body' => $post_vars );
-
-		// Make the external API call.
-		$remote_submit = wp_remote_post( self::$submit_url, $post_args );
-
-		if ( is_wp_error( $remote_submit ) ) {
-			$return['status'] = 0;
-			$return['response'] = 'Failed submitting to Liaison API. Please retry. Error: ' . $remote_submit->get_error_message();
-			error_log( sprintf( '%s: %s', __METHOD__, $return['response'] ) );
-			wp_send_json( $return );
-			return;
-		}
-
-		// Decode the response and activate redirect to the personal url on success.
-		$resp = json_decode( $remote_submit['body'] );
-
-		$return = array();
-		$return['status'] = 0;
+/**
+ * Composer autoload
+ */
+require __DIR__ . '/vendor/autoload.php';
 
-		$return['status'] 		= ( isset( $resp->status ) && 'success' == $resp->status) ? 1 : 0;
-		$return['response'] 	= ( isset( $resp->message ) ) ? $resp->message : 'Something bad happened, please refresh the page and try again.';
-		$return['data'] 		= ( isset( $resp->data ) ) ? $resp->data : '';
+$admin = new Admin();
 
-		// Return a JSON encoded reply for the validation javascript.
-		wp_send_json( $return );
-	}
+// Initialize the admin settings.
+add_action( 'admin_init', array( $admin, 'bu_liaison_inquiry_settings_init' ) );
 
-	/**
-	 * Sanitize and format post data for submission
-	 *
-	 * @param array $incoming_post_vars $_POST values as submitted.
-	 * @param array $phone_fields Array of phone field ids.
-	 * @return array Returns an array of sanitized and prepared post values.
-	 */
-	function prepare_form_post( $incoming_post_vars, $phone_fields ) {
-		// Process all of the existing values into a new array.
-		$post_vars = array();
-		foreach ( $incoming_post_vars as $key => $value ) {
-			if ( in_array( $key, $phone_fields ) ) {
-				// If it is a phone field, apply special formatting.
-				// Strip out everything except numerals.
-				$value = preg_replace( '/[^0-9]/', '', $value );
-				if ( $value ) {
-					$value = '%2B1' . $value;		// Append +1 for US, but + needs to be %2B for posting.
-				}
-			} elseif ( stripos( $key, '-text-opt-in' ) !== false ) {
-				// If this checkbox field is set then it was checked.
-				$value = '1';
-			} else {
-				// Apply basic field sanitization.
-				$value = sanitize_text_field( $value );
-			}
+// Register the page in the admin menu.
+add_action( 'admin_menu', array( $admin, 'bu_liaison_inquiry_options_page' ) );
 
-			$post_vars[ $key ] = $value;
-		}
-		return $post_vars;
-	}
-}
 
 // Instantiate plugin (only once).
 if ( ! isset( $GLOBALS['bu_liaison_inquiry'] ) ) {
-	$GLOBALS['bu_liaison_inquiry'] = new BU_Liaison_Inquiry();
+	// Get API key and Client ID from option settings.
```

Comment:
Indeed, it's done this way to simplify unit testing. When API class instance is being passed as a parameter to the Plugin class constructor, it's easier to mock its functionality.

I agree that API class shouldn't be instantiated on plugin initialization, better to do it right before it's going to be used (i.e in the shortcode handler). It will not only help to avoid the error condition you mentioned, but also will slightly improve overall performance (as we don't create an object that's most likely not used).

We can further split functionality into smaller classes to achieve this goal while keeping the ease of testing. Let's create a new class `Form_Builder` and move into it all methods of `Plugin` except the shortcode handler `liaison_inquiry_form` and the action handler `ajax_inquiry_form_post`. These two methods will instantiate both API class and Form_Builder class when needed.

Additionally, we can use [transients](https://codex.wordpress.org/Transients_API) to remind to set the options after installation of the plugin.